src/Main.elm

```diff
@@ -25,37 +68,478 @@ init =
+    | ToggleJsonMode ViewJson
 
 
 update : Msg -> Model -> ( Model, Cmd Msg )
-update _ model =
-    ( model, Cmd.none )
+update msg model =
+    ( case msg of
+        ClearEditFocus ->
+            { model | editFocus = NotEditing }
 
+        EditContactsField ( field, value ) ->
+            { model | editFocus = FocusContactsEdit ( field, value ) }
 
+        SetContactMethod pmethod ->
+            { model
+                | userGroup = Result.map (updateContactMethod pmethod) model.userGroup
+                , validationErrors =
+                    case model.userGroup of
+                        Ok userGroup ->
+                            let
+                                resValidDraftAddress =
+                                    -- just revalidating address with new contact method
+                                    CD.toDraft userGroup.contactDetails.address
+                                        |> V.validate (addressValidator pmethod)
+                            in
+                            case resValidDraftAddress of
+                                Ok _ -> []
+                                Err errors -> errors
 
----- VIEW ----
+                        Err _ ->
+                            []
+            }
+
+        TryToUpdateAddress newDraftAddress ->
+            let
+                preferredMethod =
+                    Result.map (.contactDetails >> .address >> .preferredContactMethod) model.userGroup
+                        |> Result.withDefault CD.PC_None -- if result is erronous, we won't update anything anyway
+
+                resValidDraftAddress =
+                    V.validate (addressValidator preferredMethod) newDraftAddress
+            in
+            case resValidDraftAddress of
+                Ok validDraftAddress ->
+                    { model
+                        | validationErrors = []
+                        , editFocus = NotEditing
+                        , userGroup = Result.map (updateAddress validDraftAddress) model.userGroup
+                    }
+
+                Err errors ->
+                    { model | validationErrors = errors }
+
+        SetTagInProgress inProgress ->
+            { model | editFocus = FocusTag inProgress }
+
+        TryToCreateTag { newName, newValue } ->
+            let
+                resNextUserGroup =
+                    validateTagAnd addTag (tagValidator { isNew = True, index = Nothing }) model <|
+                        Tag.make newName newValue
+            in -- The three `in` blocks here are actually the same, merge them in one action?
+            case resNextUserGroup of
+                Ok nextUserGroup ->
+                    { model
+                        | editFocus = NotEditing
+                        , validationErrors = []
+                        , userGroup = Ok nextUserGroup
+                    }
+
+                Err validationErrors ->
+                    { model | validationErrors = validationErrors }
```

Comment:
Wouldn't it be better to compute the validation errors in the `view` and keep the `Model` simpler?

Reply:
To be perfectly honest, I'm not sure I'm convinced by your argument.
A simple `getValidationErrors model` to replace `model.validationErrors` would be all the change required to the `view` function.

Feldman's did a talk about "making impossible states impossible".
Do you think the ideas described there could be applied here?


> As a ninja

What do you exactly mean by "ninja"?
I fear I'm not familiar with the term.

Reply:
> What do you exactly mean by "ninja"?

I think this refers to the anonymisation part which is manifested as the ninja emoji before each message from the candidate and they don't want to give away who they are.

Reply:
@kraklin as in, if they told me how they got to those beliefs it would give away their identity?
If this is the case, I haven't even read their resume, so that shouldn't be a problem. =)
What, is this someone I know!? O_O

Reply:
@xarvh it is some piece of information that might give calm-karl away. They don't know that you didn't read CV's on candidates so I think it is just a precaution from their side to keep the anonymity going on. I think we might just keep this topic for the call for sure then :)

Reply:
Makes sense!

Reply:
Fair enough, seems a sensible approach. =)